File: deimos_ws/src/rover_teleop/scripts/realsense_to_pointcloud.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import PointCloud, PointCloud2, ChannelFloat32, PointField
from sensor_msgs import point_cloud2
import std_msgs.msg
import geometry_msgs.msg
import numpy as np
import struct

class PointCloudConverter:
    def __init__(self):
        rospy.init_node('realsense_to_pointcloud_py', anonymous=True)
        self.cloud_pub = rospy.Publisher('cloud', PointCloud, queue_size=1)
        self.cloud_sub = rospy.Subscriber('/camera/depth/color/points', PointCloud2, self.cloud_callback)

    def cloud_callback(self, cloud_msg2):
        rospy.loginfo_once(f"Received PointCloud2 message with fields: {cloud_msg2.fields}")
        cloud_msg = PointCloud()
        cloud_msg.header = cloud_msg2.header
        cloud_msg.points = []
        cloud_msg.channels = []

        try:
            points = point_cloud2.read_points(cloud_msg2, field_names=("x", "y", "z", "rgb"), skip_nans=True)

            intensities = []
            for p in points:
                point = geometry_msgs.msg.Point32()
                point.x = p[0]
                point.y = p[1]
                point.z = p[2]
                cloud_msg.points.append(point)
                rgb_float = p[3]
                s = struct.pack('>f', rgb_float)
                i = struct.unpack('>i', s)[0]
                r = (i >> 16) & 0xFF
                g = (i >> 8) & 0xFF
                b = i & 0xFF
                intensity = (r + g + b) / 3.0
                intensities.append(intensity)

            if cloud_msg.points and intensities:
                intensity_channel = ChannelFloat32()
                intensity_channel.name = "intensities"
                intensity_channel.values = intensities
                cloud_msg.channels.append(intensity_channel)

            self.cloud_pub.publish(cloud_msg)

        except Exception as e:
            rospy.logerr(f"Error processing point cloud: {e}")

if __name__ == '__main__':
    converter = PointCloudConverter()
    rospy.spin()

File: deimos_ws/src/rover_teleop/scripts/realsense_to_pointcloud.cpp
#include <ros/ros.h>
#include <sensor_msgs/PointCloud2.h>
#include <sensor_msgs/PointCloud.h>
#include <sensor_msgs/ChannelFloat32.h>
#include <sensor_msgs/point_cloud2_iterator.h>
#include <geometry_msgs/Point32.h>

class PointCloudConverter {
public:
    PointCloudConverter() {
        ros::NodeHandle nh;
        cloud_pub_ = nh.advertise<sensor_msgs::PointCloud>("cloud", 1);
        cloud_sub_ = nh.subscribe("/camera/depth/color/points", 1, &PointCloudConverter::cloudCallback, this);
    }

private:
    ros::Publisher cloud_pub_;
    ros::Subscriber cloud_sub_;

    void cloudCallback(const sensor_msgs::PointCloud2ConstPtr& cloud_msg2) {
        static bool printed = false;
        if (!printed) {
            ROS_INFO("Received PointCloud2 message");
            printed = true;
        }

        sensor_msgs::PointCloud cloud_msg;
        cloud_msg.header = cloud_msg2->header;

        sensor_msgs::ChannelFloat32 intensity_channel;
        intensity_channel.name = "intensities";

        sensor_msgs::PointCloud2ConstIterator<float> iter_x(*cloud_msg2, "x");
        sensor_msgs::PointCloud2ConstIterator<float> iter_y(*cloud_msg2, "y");
        sensor_msgs::PointCloud2ConstIterator<float> iter_z(*cloud_msg2, "z");
        sensor_msgs::PointCloud2ConstIterator<float> iter_rgb(*cloud_msg2, "rgb");

        for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z, ++iter_rgb) {
            geometry_msgs::Point32 p;
            p.x = *iter_x;
            p.y = *iter_y;
            p.z = *iter_z;
            cloud_msg.points.push_back(p);

            // Extract RGB from packed float
            union {
                float rgb_float;
                uint32_t rgb_uint;
            } rgb_union;
            rgb_union.rgb_float = *iter_rgb;

            uint8_t r = (rgb_union.rgb_uint >> 16) & 0xFF;
            uint8_t g = (rgb_union.rgb_uint >> 8) & 0xFF;
            uint8_t b = rgb_union.rgb_uint & 0xFF;

            float intensity = (r + g + b) / 3.0f;
            intensity_channel.values.push_back(intensity);
        }

        if (!cloud_msg.points.empty() && !intensity_channel.values.empty()) {
            cloud_msg.channels.push_back(intensity_channel);
        }

        cloud_pub_.publish(cloud_msg);
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "realsense_to_pointcloud_cpp");
    PointCloudConverter converter;
    ros::spin();
    return 0;
}
